### PR TITLE
Add vector search workload with no train procedure as default

### LINKED ISSUE #140: [FEATURE]  Add new workload to test performance of knn_vector field type
### Is your feature request related to a problem?
OpenSearch KNN plugin hosts workloads  here(https://github.com/opensearch-project/k-NN/tree/main/benchmarks/osb) to run benchmarks for every release. We would like to move it to this repository so that all workloads can be in one place and  will get more visibility. 

### What solution would you like?
Create a new workload to run workloads using knn_vector to measure vector search performance.

### What alternatives have you considered?
N/A

### Do you have any additional context?
N/A

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
Add vector search workload to benchmark performance of indexing and search using knn_vector as field type.

### Issues Resolved
#140 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
This PR contains benchmark workload that was previously added to [knn repository ](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/osb). This PR contains only indexing and search component. Other features like training will be added in subsequent PR.

**Comment by VijayanB** [comment]
@rishabh6788 To run this workload. we have dependencies to library like numpy and h5py. Should this be added in this workload or to opensearch-benchmark repository?

File: knnvector/test_procedures/default.json
Comment (navneet1v) [inline knnvector/test_procedures/default.json:58]: this will not necessary load all the graphs in memory. The graphs from replicas may not be loaded.

Comment (VijayanB) [reply knnvector/test_procedures/default.json:58]: Got it. Should i change it to warmup API then?

Comment (navneet1v) [reply knnvector/test_procedures/default.json:58]: I would like to have both.

Comment (navneet1v) [reply knnvector/test_procedures/default.json:58]: Yes we should remove this and move this to warmup api call.

Also can we make the warmup step optional, where by default it is true

File: knnvector/README.md
Comment (navneet1v) [inline knnvector/README.md:3]: Lets change this to something, This workload is to benchmark performance of indexing and search of Vector Engine of Opensearch

File: knnvector/operations/default.json
Comment (navneet1v) [inline knnvector/operations/default.json:21]: lets use a default value as 1 for this.

File: knnvector/params/nmslib-sift-128-l2.json
Comment (navneet1v) [inline knnvector/params/nmslib-sift-128-l2.json:11]: is this path correct?

File: knnvector/params_sources.py
Comment (navneet1v) [inline knnvector/params_sources.py:136]: This logic has potential to index more documents as compared to what is provided. If we have plans to fix this in next PR lets add a comment on top of this to ensure that we are not missing fixing the bug.

File: knnvector/runners.py
Comment (navneet1v) [inline knnvector/runners.py:41]: why do we need this custom operation for ingestion? can't we use the Bulk ingestion operation provided in OSB?

File: knnvector/runners.py
Comment (navneet1v) [inline knnvector/runners.py:64]: same as above.

File: knnvector/README.md
Comment (navneet1v) [inline knnvector/README.md:5]: I don't see any sample commands to run the workload. Lets make sure that we add some sample commands and format on how to run the workload.

File: knnvector/operations/default.json
Comment (rishabh6788) [reply knnvector/operations/default.json:21]: What's the reasoning behind defaulting to 1 segment? Does that give better performance result?

Comment (navneet1v) [reply knnvector/operations/default.json:21]: yes, single segment leads to best performance for the vector engine as segments are searched sequentially.

File: knnvector/runners.py
Comment (VijayanB) [reply knnvector/runners.py:41]: This was introduced to make sure that no documents are left out due to connection issue. This was observed during our testing. Since, recall assumes all documents are indexed completely, we added it originally.  I will add TODO to address this issue.

File: knnvector/README.md
Comment (VijayanB) [reply knnvector/README.md:5]: This is just workload repository. OSB invokes this repository, and i believe dependencies should be added to OSB. i asked this question here https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/144#issuecomment-1823675797 , waiting for response.

File: knnvector/README.md
Comment (VijayanB) [reply knnvector/README.md:3]: Do you think we should change workload name from knn vector to Vector engine too?

File: knnvector/params/nmslib-sift-128-l2.json
Comment (VijayanB) [reply knnvector/params/nmslib-sift-128-l2.json:11]: This path should be set by users. In future we can simplify by using corpora. For now, will keep the path as user defined.

File: knnvector/params_sources.py
Comment (VijayanB) [reply knnvector/params_sources.py:136]: Ack

File: knnvector/runners.py
Comment (VijayanB) [reply knnvector/runners.py:64]: Removed this class and used Refresh from OSB

File: knnvector/operations/default.json
Comment (VijayanB) [reply knnvector/operations/default.json]: Ack

File: knnvector/params_sources.py
Comment (jmazanec15) [inline knnvector/params_sources.py:157]: Can you add repetitions here?

File: knnvector/params_sources.py
Comment (jmazanec15) [inline knnvector/params_sources.py:159]: I wonder if this is even necessary. We could probably remove it at a certain point. A disk read will read the 4096 bytes into the page cache anyway (I believe). Maybe for another review

File: knnvector/runners.py
Comment (jmazanec15) [inline knnvector/runners.py:24]: Could you add reasoning why we dont need this? I believe its because we should not need retry logic

File: knnvector/test_procedures/default.json
Comment (jmazanec15) [inline knnvector/test_procedures/default.json:46]: Do we need this if we are forcemerging?

File: knnvector/params_sources.py
Comment (VijayanB) [reply knnvector/params_sources.py:157]: Ack

File: knnvector/params_sources.py
Comment (VijayanB) [reply knnvector/params_sources.py:159]: Sure. I will remove it in next PR.

File: knnvector/runners.py
Comment (VijayanB) [reply knnvector/runners.py:24]: Sure.

File: knnvector/test_procedures/default.json
Comment (VijayanB) [reply knnvector/test_procedures/default.json:35]: Moved refresh above force-merge

**Review by jmazanec15 — APPROVED** [review]
LGTM!

File: knnvector/params/nmslib-sift-128-l2.json
Comment (rishabh6788) [reply knnvector/params/nmslib-sift-128-l2.json:11]: Does this mean the user needs to download the dataset manually, put it in the desired path and then declare it here?

File: knnvector/test_procedures/default.json
Comment (IanHoang) [inline knnvector/test_procedures/default.json:3]: To conform with other official workloads and since this is the default test procedure, it'd be helpful to add a "description" field with a blurb about what this test procedure does.

File: knnvector/README.md
Comment (IanHoang) [inline knnvector/README.md:145]: Nit: `to to` --> `to`

File: knnvector/test_procedures/default.json
Comment (VijayanB) [reply knnvector/test_procedures/default.json:4]: Ack

File: knnvector/params/nmslib-sift-128-l2.json
Comment (VijayanB) [reply knnvector/params/nmslib-sift-128-l2.json:11]: Yes. I am exploring how to download default/popular dataset using corpora. This can be added incremental

File: knnvector/README.md
Comment (VijayanB) [reply knnvector/README.md]: Ack

**Review by jmazanec15 — APPROVED** [review]

File: knnvector/README.md
Comment (VijayanB) [reply knnvector/README.md:5]: Moved all dependencies to opensearch-benchmark repository and added section on how to run osb

**Review by IanHoang — CHANGES_REQUESTED** [review]
There might be users who are unfamiliar with `faiss`, `nmslib`, or characteristics of KNN workloads. 

Could you elaborate in the README what those are and how they should be used? 

Also, what's the significance behind the `sift-128-l2` in the the params files names? Could you provide some insight into that as well in the README if they're important?

**Review by IanHoang — COMMENTED** [review]
Have we decided on a workload name? Will it be `knn-workload` or `texmex`? The former might be too generic. Could be the latter since workload is based off of the dataset from that link.

File: knnvector/README.md
Comment (IanHoang) [inline knnvector/README.md:36]: This should be updated to `--workload=<WORKLOAD_NAME>`

**Comment by VijayanB** [comment]
Texmex is an example, this is similar to how we have workloads like geopoint / geoshape. This workload is going to be based on knn-vector field type. IMO, we can call this either knnvector or vectorsearch.

File: knnvector/README.md
Comment (VijayanB) [reply knnvector/README.md:36]: Do you recommend to replace path or append this parameter?

Comment (IanHoang) [reply knnvector/README.md:36]: `--workload` takes in the name of the workload or the folder name of this workload (`knnvector`)

**Comment by IanHoang** [comment]
For users not familiar with 

> Texmex is an example, this is similar to how we have workloads like geopoint / geoshape. This workload is going to be based on knn-vector field type. IMO, we can call this either knnvector or vectorsearch.

Understood. I see the folder name is named `knnvector`. We can opt to keep it as `knnvector` then

File: knnvector/README.md
Comment (IanHoang) [reply knnvector/README.md:36]: We should replace `--workload-path` and use `--workload`. However, `--workload` takes in the directory name or workload name so it will look like `--workload knnvector`

**Comment by navneet1v** [comment]
> Texmex is an example, this is similar to how we have workloads like geopoint / geoshape. This workload is going to be based on knn-vector field type. IMO, we can call this either knnvector or vectorsearch.

Lets use vectorsearch. no mention of kNN should be there.

**Comment by VijayanB** [comment]
> There might be users who are unfamiliar with `faiss`, `nmslib`, or characteristics of KNN workloads.
> 
> Could you elaborate in the README what those are and how they should be used?
> 
> Also, what's the significance behind the `sift-128-l2` in the the params files names? Could you provide some insight into that as well in the README if they're important?

Those are part of index mapping definition which is not fixed. I will add line about where to find more details.

**Review by IanHoang — APPROVED** [review]

File: vectorsearch/README.md
Comment (IanHoang) [inline vectorsearch/README.md:23]: We're missing the params file `no-trains-params.json`. 
image

It might be more clear to say something like this:  
```
To run the test procedure, create a file based off of the contents of either `faiss-sift-128-l2.json`or `nmslib-sift-128-l2.json` and modify the parameters. Once parameters are set in the file, export the URL, PORT, TEST PROCEDURE, and params file you just created. Run the following command to run the no trains test procedure
```

**Review by IanHoang — CHANGES_REQUESTED** [review]
Left a few comments.

File: vectorsearch/README.md
Comment (IanHoang) [inline vectorsearch/README.md:36]: should be `--workload vectorsearch \`

File: vectorsearch/README.md
Comment (IanHoang) [inline vectorsearch/README.md:38]: Since `no-trains-test` is the only test procedure and default, no need to specify

Comment (VijayanB) [reply vectorsearch/README.md]: Ack

**Comment by IanHoang** [comment]
> There is no coupling between params file name and workload. I named it such a way that it gives hint on selected param values .

@VijayanB I understand now. To clarify this, could you add a section in the README stating that the files `faiss-sift-128-l2` and `nmslib-sift-128-l2` are sample params that can be used in the workload and can be used as reference params file for users who want to make their own custom params file.

**Comment by VijayanB** [comment]
> > There is no coupling between params file name and workload. I named it such a way that it gives hint on selected param values .
> 
> @VijayanB I understand now. To clarify this, could you add a section in the README stating that the files `faiss-sift-128-l2` and `nmslib-sift-128-l2` are sample params that can be used in the workload and can be used as reference params file for users who want to make their own custom params file. EDIT: See that it was added now 👍🏻

@IanHoang Any pending comments needs to be addressed?

**Review by IanHoang — APPROVED** [review]

**Comment by VijayanB** [comment]
@gkamat can you take a look at this PR? Thanks

File: vectorsearch/workload.json
Comment (gkamat) [inline vectorsearch/workload.json:4]: A more verbose description will be helpful for users not familiar with the domain here.

**Review by gkamat — CHANGES_REQUESTED** [review]

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:18]: @gkamat I updated the text. We will add new procedure which can use model . Do you recommend to mention this future work in README?

File: vectorsearch/indices/faiss-index.json
Comment (VijayanB) [reply vectorsearch/indices/faiss-index.json:21]: it is required parameter

File: vectorsearch/test_procedures/default.json
Comment (VijayanB) [reply vectorsearch/test_procedures/default.json:33]: We will have another type of index called training index for training models. We kept target index to differentiate between the indexing storing the training data and the index that will contain the actual data

File: vectorsearch/workload.json
Comment (VijayanB) [reply vectorsearch/workload.json:4]: Ack

File: vectorsearch/test_procedures/default.json
Comment (VijayanB) [reply vectorsearch/test_procedures/default.json:3]: Ack

File: vectorsearch/runners.py
Comment (VijayanB) [reply vectorsearch/runners.py:40]: Ack

File: vectorsearch/runners.py
Comment (VijayanB) [reply vectorsearch/runners.py:38]: Ack

File: vectorsearch/params/faiss-sift-128-l2.json
Comment (VijayanB) [reply vectorsearch/params/faiss-sift-128-l2.json:3]: We will have another type of index called model index for training models. We kept target index to differentiate between the indexing storing the training data and the index that will contain the actual data

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:81]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:172]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:37]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:24]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:22]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:20]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:14]: Ack

File: vectorsearch/README.md
Comment (VijayanB) [reply vectorsearch/README.md:7]: Ack

File: vectorsearch/indices/faiss-index.json
Comment (gkamat) [reply vectorsearch/indices/faiss-index.json:21]: Ack.

File: vectorsearch/test_procedures/default.json
Comment (gkamat) [reply vectorsearch/test_procedures/default.json]: Ack.

**Comment by gkamat** [comment]
Please confirm this is intended for backport to both the 1 and 2 branches.

File: vectorsearch/README.md
Comment (gkamat) [reply vectorsearch/README.md]: Yes, that would be ideal.  Subsequently, you can update the writeup when the new procedure gets added.

**Review by gkamat — APPROVED** [review]
Please confirm the backport labels are set correctly before merging.  Thanks.

**Comment by VijayanB** [comment]
@gkamat yes, This is supported for both OpenSearch 1.x and 2.x

**Pull request merged**
